### ISSUE TITLE
Functions to escape Kamelet parameters

### ISSUE THREAD

**Issue by nicolaferraro** [apache/camel-k-runtime#632]
There are multiple cases where I want to escape a specific parameter. But "escaping", depending on the context, could mean different things.

Suppose I've a Kamelet with a `param`.

If the `param` represents an ID in a rest resource, I could use it as:

```yaml
- tod: "https://api.com/users/{{param}}/details"
```

If I pass `luca` it works. If I pass `luca/tokens?` I did an injection attack.

Here I'd like to URL encode the param as URL path segment.

In another case, I need to analyze the output of a service to call another one, so I use JSONpath:

```
- to: https://api.com/endpoint
- transform:
    jsonpath: "$.value[?(@.lastName == '{{param}}')]"
```

If I pass `Burgazzoli` it works. If I pass `O'Neil` then the whole integration crashes because of invalid JSONpath expression.

How can we solve thes issues?

**Comment by lburgazzoli** [comment]
Guess the same happens on plain camel right ?

**Comment by davsclaus** [comment]
Yes the O Neil thing would be same problem. Can you escape the quote or use double quotes for the == {{param}}

**Comment by davsclaus** [comment]
Maybe a kamelet parameter needs to a way to specify valid inputs. So you can say that valid is only alpha numeric, or max length is 10, min lenght is 5, etc. And other kinds of rules.

For O Niel then simple language allows escape, so you can say O\'Niel

**Comment by nicolaferraro** [comment]
> Maybe a kamelet parameter needs to a way to specify valid inputs. So you can say that valid is only alpha numeric, or max length is 10, min lenght is 5, etc. And other kinds of rules.
> 
> For O Niel then simple language allows escape, so you can say
> 
> ```
>     O\'Niel
> ```
> 
> But this is jsonpath in the example

Yeah, that's an option, but not the more user-friendly in the general case, since it would require that the user manually escapes `O'Neil` into `O\'Neil`. Ideally it's the template mechanism that should allow the template writer to apply escaping functions in order to use the "transformed" version of the parameter in different parts of the route template... But I don't know how, or if it could be done declaratively.

**Comment by davsclaus** [comment]
I created a ticket about this problem as its a general camel-core problem
https://issues.apache.org/jira/browse/CAMEL-16242

**Comment by davsclaus** [comment]
Okay so the single quoutes should be double quotes, then we can use O'Niel, such as

                        .transform().jsonpath("$.store.book[?(@.author == \"John O'Niel\")].title")

Since kamelets in yaml dsl are already using double quotes for string literals. We can maybe transpose the expression by looking for single quote pairs, and replace them as double quotes, for specific languages like simple or jsonpath.

**Comment by davsclaus** [comment]
Another alternative is to see if the placeholder is used where there are single quotes before / after, and then escape internal single quotes

**Comment by lburgazzoli** [comment]
> Since kamelets in yaml dsl are already using double quotes for string literals. We can maybe transpose the expression by looking for single quote pairs, and replace them as double quotes, for specific languages like simple or jsonpath.

I think in YAML you can use single quotes instead of double quotes

**Comment by davsclaus** [comment]
Okay so the kamelet yaml could be

```
- to: https://api.com/endpoint
- transform:
    jsonpath: '$.value[?(@.lastName == "{{param}}")]'
```

**Comment by davsclaus** [comment]
JIRA about to configure input rules
https://issues.apache.org/jira/browse/CAMEL-16312

**Comment by davsclaus** [comment]
This is fixed in Camel 3.10, so you can use whatever syntax in the kamelets you like
